### PR TITLE
Re-enable MacOS signing tests

### PR DESCRIPTION
**Pull request by PatoBeltran**
## Bug

Fixes: https://github.com/NuGet/Home/issues/7608
## Fix

Re-enabled and made pass mac tests. This change involves changing some utilities to be able to create certificates with private keys cross platform. Test utility had to be moved to netcoreapp2.1 since netstandard2.0 doesn't have support for the methods needed. Once netstandard2.1 is out test.utility should be able to return to target netstandard instead of netcore.

### CODE REVIEW

File: src/NuGet.Core/NuGet.Commands/SignCommand/CertificateProvider.cs [hunk 1]
```diff
@@ -50,38 +54,53 @@ public static async Task<X509Certificate2Collection> GetCertificatesAsync(Certif
 
                     resultCollection = new X509Certificate2Collection(cert);
                 }
-                catch (CryptographicException ex)
+                catch (Exception ex)
                 {
-                    switch (ex.HResult)
+                    if (ex is CryptographicException ce)
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/SignCommand/CertificateProvider.cs]: You don't need the if statement. 

```cs
catch(CryptographicException ce)
{
// code
} 
catch (FileNotFoundException fne) 
{
// code
}
```

File: test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/CertificateChainUtilityTests.cs [hunk 2]
```diff
@@ -127,11 +131,11 @@ public void GetCertificateChain_WithUntrustedSelfIssuedCertificate_ReturnsChain(
                 }
 
                 Assert.Equal(0, logger.Errors);
-                Assert.Equal(RuntimeEnvironmentHelper.IsWindows ? 1 : 2, logger.Warnings);
+                Assert.Equal(RuntimeEnvironmentHelper.IsLinux ? 2 : 1, logger.Warnings);
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/CertificateChainUtilityTests.cs:134]: what's the extra linux warning?

**Review by nkolev92 — APPROVED** [review]
LGTM.

Make sure all the tests are passing before merging.

File: test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/CertificateChainUtilityTests.cs (hunk 2 above)
Comment (PatoBeltran) [reply test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/CertificateChainUtilityTests.cs:134]: Since the certificate is self-issued, Linux also includes `"unable to get certificate CRL"`, which is a valid warning message to give (since self issued certificates are usually not part of a CRL)

File: test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/TimestampProviderTests.cs [hunk 3]
```diff
@@ -139,10 +137,10 @@ public async Task GetTimestamp_WhenRequestNull_ThrowsAsync()
             var authorCertName = "author@nuget.func.test";
             var content = new SignatureContent(SigningSpecifications.V1, Common.HashAlgorithmName.SHA256, "Test data to be signed and timestamped");
 
-            Action<X509V3CertificateGenerator> modifyGenerator = delegate (X509V3CertificateGenerator gen)
+            Action<TestCertificateGenerator> modifyGenerator = delegate (TestCertificateGenerator gen)
             {
-                gen.SetNotBefore(DateTime.MinValue);
-                gen.SetNotBefore(DateTime.UtcNow.Subtract(TimeSpan.FromDays(1))); // cert has expired
+                gen.NotBefore = DateTime.MinValue;
```
Comment (dtivel) [inline test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/TimestampProviderTests.cs]: Is this line necessary?  It looks like the value is being overwritten by the next statement.

File: test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/CertificateUtilityTests.cs [hunk 4]
```diff
@@ -223,10 +233,12 @@ public void IsValidForPurposeFast_WithAnyExtendedKeyUsage_ReturnsFalseForCodeSig
...
-                    generator.AddExtension(
-                        X509Extensions.ExtendedKeyUsage.Id,
-                        critical: false,
-                        extensionValue: new ExtendedKeyUsage(KeyPurposeID.IdKPEmailProtection, KeyPurposeID.AnyExtendedKeyUsage));
+                    var usages = new OidCollection { TestCertificateGenerator.IdKPEmailProtection, TestCertificateGenerator.AnyExtendedKeyUsage };
+
+                    generator.Extensions.Add(
+                        new X509EnhancedKeyUsageExtension(
+                              usages,
+                              critical: true));
```
Comment (dtivel) [inline test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/CertificateUtilityTests.cs]: `critical` should be `false`.  From https://tools.ietf.org/html/rfc5280#section-4.2.1.12:
>If a CA includes extended key usages to satisfy such applications,
   but does not wish to restrict usages of the key, the CA can include
   the special KeyPurposeId anyExtendedKeyUsage in addition to the
   particular key purposes required by the applications.  Conforming CAs
   SHOULD NOT mark this extension as critical if the anyExtendedKeyUsage
   KeyPurposeId is present.

File: test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/CertificateUtilityTests.cs [hunk 5]
```diff
@@ -162,10 +166,12 @@ public void HasExtendedKeyUsage_WithEku_ReturnsTrueForEku()
...
-                    generator.AddExtension(
-                        X509Extensions.ExtendedKeyUsage.Id,
-                        critical: false,
-                        extensionValue: new ExtendedKeyUsage(KeyPurposeID.IdKPCodeSigning));
+                    var usages = new OidCollection { TestCertificateGenerator.IdKPCodeSigning };
+
+                    generator.Extensions.Add(
+                        new X509EnhancedKeyUsageExtension(
+                              usages,
+                              critical: true));
```
Comment (dtivel) [inline test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/CertificateUtilityTests.cs]: With the exception of the test [`IsValidForPurposeFast_WithAnyExtendedKeyUsage_ReturnsFalseForCodeSigningEku`](https://github.com/NuGet/NuGet.Client/pull/2685/files#diff-1267bcc195ace161677a111cfd7b09cbR231) below, why did you change `critical` from `false` to `true` for these tests?  I don't think it's a big deal either way --- just curious.

File: test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/CertificateUtilityTests.cs [hunk 6]
```diff
@@ -26,22 +28,23 @@ public CertificateUtilityTests(CertificatesFixture fixture)
         }
 
         [Theory]
-        [InlineData("SHA256WITHRSAENCRYPTION", Oids.Sha256WithRSAEncryption)]
-        [InlineData("SHA384WITHRSAENCRYPTION", Oids.Sha384WithRSAEncryption)]
-        [InlineData("SHA512WITHRSAENCRYPTION", Oids.Sha512WithRSAEncryption)]
-        public void IsSignatureAlgorithmSupported_WhenSupported_ReturnsTrue(string signatureAlgorithm, string expectedSignatureAlgorithmOid)
+        [InlineData(Common.HashAlgorithmName.SHA256, RSASignaturePaddingMode.Pkcs1, Oids.Sha256WithRSAEncryption)]
```
Comment (dtivel) [inline test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/CertificateUtilityTests.cs]: `RSASignaturePaddingMode.Pkcs1` could be removed as a parameter and inlined within the test, since it never varies.

File: src/NuGet.Core/NuGet.Commands/SignCommand/CertificateProvider.cs [hunk 7]
```diff
@@ -83,6 +89,14 @@ public static async Task<X509Certificate2Collection> GetCertificatesAsync(Certif
                             throw;
                     }
                 }
+                catch (FileNotFoundException)
+                {
+                        throw new SignCommandException(
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Commands/SignCommand/CertificateProvider.cs]: Indentation looks off.

File: test/TestUtilities/Test.Utility/Signing/TestCertificateGenerator.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+using System;
+using System.Collections.ObjectModel;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+
+namespace Test.Utility.Signing
+{
+    public class TestCertificateGenerator
+    {
+        public static readonly Oid IdKPCodeSigning = new Oid("1.3.6.1.5.5.7.3.3");
```
Comment (dtivel) [inline test/TestUtilities/Test.Utility/Signing/TestCertificateGenerator.cs]: Test OID's should go in their own static class like we do [here](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Packaging/Signing/Oids.cs).

Also, for code signing you can use `NuGet.Packaging.Signing.Oids.CodeSigningEku` instead of redefining it here.

Also, you're not obliged to keep BouncyCastle names.  I'd change `IdKP` to `IdKp`.

File: test/TestUtilities/Test.Utility/Signing/TestCertificateGenerator.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        public static readonly Oid IdKPClientAuth = new Oid("1.3.6.1.5.5.7.3.2");
+        public static readonly Oid IdKPEmailProtection = new Oid("1.3.6.1.5.5.7.3.4");
+        public static readonly Oid AnyExtendedKeyUsage = new Oid("2.5.29.37");
```
Comment (dtivel) [inline test/TestUtilities/Test.Utility/Signing/TestCertificateGenerator.cs]: This is wrong.  It's 2.5.29.37.0.

Search [RFC 5280](https://tools.ietf.org/html/rfc5280) for `anyExtendedKeyUsage`.

Also see http://oid-info.com/cgi-bin/display?oid=2.5.29.37.0&action=display

File: test/TestUtilities/Test.Utility/Signing/TestCertificateGenerator.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.ObjectModel;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+
+namespace Test.Utility.Signing
+{
+    public class TestCertificateGenerator
```
Comment (dtivel) [inline test/TestUtilities/Test.Utility/Signing/TestCertificateGenerator.cs]: `sealed`?

File: test/TestUtilities/Test.Utility/Signing/X509CertificateWithKeyInfo.cs [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+
+namespace Test.Utility.Signing
+{
+    /// <summary>
+    /// Utility to store a certificate and the RSA key that it was build with.
```
Comment (dtivel) [inline test/TestUtilities/Test.Utility/Signing/X509CertificateWithKeyInfo.cs]: `built`

File: test/TestUtilities/Test.Utility/Signing/X509CertificateWithKeyInfo.cs [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    /// </summary>
+    /// <remarks>
+    /// This is useful because CGN certs import the key as Exportable, but bouncy castle needs it as ExportablePlainText.
```
Comment (dtivel) [inline test/TestUtilities/Test.Utility/Signing/X509CertificateWithKeyInfo.cs]: `CNG`

File: test/TestUtilities/Test.Utility/Signing/X509CertificateWithKeyInfo.cs [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    /// By having the RSA key stored, you can use it with bouncy castle. If we update our test infrastructure to not use bouncy castle
+    /// this class won't be needed anymore.
+    /// </remarks>
+    public class X509CertificateWithKeyInfo : IDisposable
```
Comment (dtivel) [inline test/TestUtilities/Test.Utility/Signing/X509CertificateWithKeyInfo.cs]: `sealed`

**Review by dtivel — APPROVED** [review]

File: test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/CertificateUtilityTests.cs (hunk 5 above)
Comment (PatoBeltran) [reply test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/CertificateUtilityTests.cs]: This wasn't on purpose. I will revert it to minimize the changes.
